feat: freight creation criteria expressions (#5047)
akuity/kargo: docs/docs/50-user-guide/60-reference-docs/40-expressions.md

features.
:::

> ### `commitFrom`

Reviewer: ```suggestion
### `commitFrom()`
```

> The `commitFrom` function signature varies based on where it's being used.

Reviewer: Just so it sounds less scary, think we should work in "slightly":

```suggestion
The signature and return value of the `commitFrom()` function vary slightly
with the context in which it's used.
```

> In the context of `DiscoveredArtifacts`, the function signature is:

Reviewer: This might not be well understood. Its meaning would probably be more clear if we talked about "Freight creation criteria":

```suggestion
In the context of an optional expression evaluated to determine whether criteria
have been met for automatic creation of a `Freight` resource following
`Warehouse`'s artifact discovery process, the function signature is:
```

`commitFrom(repoURL)`

> It has one required argument.

Reviewer: Seems to be how we did this in other places:

```suggestion
It has one required argument:
```

  commitID: ${{ commitFrom("https://github.com/example/repo.git", warehouse("my-warehouse")).ID }}
```

### `imageFrom`

The `imageFrom` function signature varies based on where it's being used.

> In the context of `DiscoveredArtifacts`, the function signature is:

Reviewer: ```suggestion
In the context of an optional expression evaluated to determine whether criteria
have been met for automatic creation of a `Freight` resource following a
`Warehouse`'s artifact discovery process, the function signature is:
```

| `Digest` | The digest of the image. |
| `Annotations` | A map of [annotations](https://specs.opencontainers.org/image-spec/annotations/) discovered for the image. |

Example:

```yaml
spec:
  freightCreationPolicy: Automatic
  subscriptions:
  - image:
      repoURL: ghcr.io/example/frontend
  - image:
      repoURL: ghcr.io/example/backend
  freightCreationCriteria:
  - expression: |
  commitFrom('ghcr.io/example/frontend').tag == 
  comitFrom('ghcr.io/example/backend').tag
```

:::info

> The returned `Image` object is the same for all contexts. Regardless of whether its used for
> `DiscoveredArtifacts`, `Promotion`, `Stage`, or `FreightCollection` resources.

Reviewer: ```suggestion
The returned `Image` object is the same in all contexts.
```

  imageTag: ${{ imageFrom("public.ecr.aws/nginx/nginx", warehouse("my-warehouse")).Tag }}
```

### `chartFrom`

> The `chartFrom` function signature varies based on where it's being used.

Reviewer: ```suggestion
The signature and return value of the `chartFrom()` function vary slightly
with the context in which it's used.
```

> In the context of `DiscoveredArtifacts`, the function signature is:

Reviewer: ```suggestion
In the context of an optional expression evaluated to determine whether criteria
have been met for automatic creation of a `Freight` resource following a
`Warehouse`'s artifact discovery process, the function signature is:
```

`chartFrom(repoURL, [chartName])`

It has one required and one optional argument:

- `repoURL` (Required): The URL of a Helm chart repository.
- `chartName` (Optional): The name of the chart (required for HTTP/S
  repositories, not needed for OCI registries).

The `chartFrom()` function returns a corresponding `Chart` object.

| Field | Description |
|-------|-------------|
| `RepoURL` | The URL of the Helm chart repository the chart originates from. For HTTP/S repositories, this is the URL of the repository. For OCI repositories, this is the URL of the container image repository including the chart's name. |
| `Name` | The name of the Helm chart. Only present for HTTP/S repositories. |
| `Version` | The version of the Helm chart. |

Example:

```yaml
spec:
  freightCreationPolicy: Automatic
  subscriptions:
  - chart:
      repoURL: oci://example.com/my-chart
  - chart:
      repoURL: oci://example.com/my-other-chart
  freightCreationCriteria:
  - expression: |
  chartFrom('oci://example.com/my-chart').version == 
  chartFrom('oci://example.com/my-other-chart').version
```

:::info

> The returned `Chart` object is the same for all contexts. Regardless of whether its used for
> `DiscoveredArtifacts`, `Promotion`, `Stage`, or `FreightCollection` resources.

Reviewer: ```suggestion
The returned `Chart` object is the same in all contexts.
```

:::

> In all other contexts such as `Promotion`, `Stage`, or `FreightCollection`, 
> it has one required and two
> optional arguments:

Reviewer: ```suggestion
In all other contexts, such as promotion and verification processes, the
function signature is:
```

`chartFrom(repoURL, [chartName], [freightOrigin])`

> 

Reviewer: ```suggestion

It has one required and two optional arguments:

```
